### ISSUE TITLE
Batch norm layer causes Horovod to hang

### ISSUE THREAD

**Issue by andrewor14** [horovod/horovod#1271]
**Environment:**
1. Framework: TF
2. Framework version: TF2.0 beta
3. Horovod version: master branch, 958695e7343ce470ad3b0d9df1967b5af3bd6ec3 (7/10/19)
4. MPI version: 4.0.1
5. CUDA version: N/A
6. NCCL version: N/A
7. Python version: 3.6
8. OS and version: Springdale (Red Hat)
9. GCC version: 4.8.5

**Your question:**
Hello, I am trying to run Horovod with the [models](https://github.com/tensorflow/models) repo. In particular, I am running cifar10 with [resnet56 using the keras API](https://github.com/tensorflow/models/blob/master/official/resnet/keras/resnet_cifar_model.py). I am not using GPUs or the distribution strategy API in tensorflow.

There is a mismatch among gradient tensor names across the cluster. For example, worker 1 sees
```
Gradient: SGD/gradients/bn4block_6_branch2b/cond_grad/Identity_1:0
Gradient: SGD/gradients/bn4block_6_branch2b/cond_grad/Identity_2:0
```
while worker 2 sees
```
Gradient: SGD/gradients/bn4block_6_branch2b/cond_grad/Identity:0
Gradient: SGD/gradients/bn4block_6_branch2b/cond_grad/Identity_1:0
```
and I get a bunch of messages such as the following:
```
One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock.
Stalled ranks:
0: [training/SGD_Allreduce/HorovodAllreduce_SGD_gradients_bn2block_2_branch2a_cond_grad_Identity_0 ...]
1: [training/SGD_Allreduce/HorovodAllreduce_SGD_gradients_bn2block_5_branch2a_cond_grad_Identity_0 ...]
```
I know it's caused by the batch norm layers because training proceeds as normal if I remove them, and I no longer get the above warnings.

How can I fix this?

**Comment by Himscipy** [comment]
@andrewor14 Were you able to find a solution for this   ? I am facing similar issue with the resnet model arch. Thank you

**Comment by thecooltechguy** [comment]
I'm facing the same issue, did anyone happen to find a fix? @Himscipy @andrewor14

**Comment by Himscipy** [comment]
@thecooltechguy  For me the new version of Horovod- 0.18.1 solved the issue. You can all play with stall time settings which you can change if that helps.

**Comment by thecooltechguy** [comment]
Thanks for the reply @Himscipy ! I was using Horovod 0.19.1 and had this issue, so I will try upgrading to 0.19.2 or downgrading to 0.18.1 to see which solves the issue.

**Comment by thecooltechguy** [comment]
For some reason, @Himscipy 's suggestion didn't work for me when I downgraded to 0.18.2 (or when I upgraded to 0.19.2). I'm still having the stalled tensors errors for the batch normalization layers, and this only goes away if I call hvd.load_model on all horovod workers (instead of on just rank 0 and then using BroadcastVariables hook to sync them with the other workers). However, if I call hvd.load_model on all workers, the loss doesn't seem to decrease thereafter (while it was decreasing before that checkpoint steadily). Is calling hvd.load_model on all workers the right approach @Himscipy ? Thanks for your help!

**Comment by Himscipy** [comment]
@thecooltechguy  Let me know if you can share a demo code to reproduce  your error. Some details about the model size,  network structure  and number of parameter would further help in understanding your case.

**Comment by thecooltechguy** [comment]
@Himscipy It's basically a ResNet-based CNN with 7.1M parameters being trained on a ~1M audio file dataset. It uses a lot of Batch Norm for the input into the CNN and within the Conv blocks (expected to be trained for a few hundred epochs).

I, unfortunately, can't share the exact code but in your experience, do you call hvd.load_model(checkpoint_file) on each rank when resuming from a checkpoint (which is what "works" for me but the loss increases once I resume, indicating that something about the weights/optimizer is not optimal when loading the model on all ranks), or do you only hvd.load_model on rank 0 and then just build/compile the model on the other ranks? I saw some mentions online that option 2 is the typical approach with horovod but that never seems to work for me due to the BatchNorm layers. I'm also using TF 2.1.0.

**Comment by thecooltechguy** [comment]
@Himscipy Actually, is it okay if you can share the TF version that you used? I've been using TF 2.1.0.
